### ISSUE TITLE
[java] New Rule: Public and protected static fields must be final

### ISSUE THREAD

**Issue by Fernal73** [pmd/pmd#2987]
**Proposed Rule Name:**
~StaticFieldMustBeFinal~ MutableStaticState

**Proposed Category:** 
~Security~ Design

**Description:**

Public and protected static fields must be made constants (or immutable references) by declaring them final.

https://rules.sonarsource.com/java/RSPEC-1444
'There is no good reason to declare a field "public" and "static" without also declaring it "final". Most of the time this is a kludge to share a state among several objects. But with this approach, any object can do whatever it wants with the shared state, such as setting it to null.'

https://www.oracle.com/java/technologies/javase/seccodeguide.html#6-9

**Guideline 6-9 / MUTABLE-9: Make public static fields final**

"Callers can trivially access and modify public non-final static fields. Neither accesses nor modifications can be guarded against, and newly set values cannot be validated. Fields with subclassable types may be set to objects with malicious implementations. Always declare public static fields as final.
....
Protected static fields suffer from the same problem as their public equivalents but also tend to indicate confused design."

**Code Sample:**

**Noncompliant Code Example**
```java
public class Greeter { public static Foo foo = new Foo(); ... } 
```

**Compliant Solution**
```java
public class Greeter { public static final Foo FOO = new Foo(); ...
```

**Possible Properties:**
None

*   Should this rule be customizable via properties?
No. See above.

**Possibly affected rules**:
[AssignmentToNonFinalStatic](https://pmd.github.io/latest/pmd_rules_java_errorprone.html#assignmenttononfinalstatic)
